File: src/behavioral/visitor3.java
public interface Visitor {
    void visitBank(Bank bank);
    void visitRestaurant(Restaurant restaurant);
    void visitResident(Resident resident);
}

@Data
public abstract class Client {
    private final String name;
    private final String address;
    private final String number;

    public abstract void accept(Visitor visitor);
}

@Getter
public class Bank extends Client {
    ...

    @Override
    public void accept(Visitor visitor) {
        visitor.visitBank(this);
    }
}

@Getter
public class Restaurant extends Client {
    ...

    @Override
    public void accept(Visitor visitor) {
        visitor.visitRestaurant(this);
    }
}

@Getter
public class Resident extends Client {
    ...

    @Override
    public void accept(Visitor visitor) {
        visitor.visitResident(this);
    }
}

public class InsuranceMessagingVisitor implements Visitor {
    public void sendInsuranceMails(List<Client> clients) {
        for (Client client : clients) {
            client.accept(this);
        }
    }

    ...
}

File: src/behavioral/visitor3.ts
interface Visitor {
  visitBank(bank: Bank): void;
  visitRestaurant(restaurant: Restaurant): void;
  visitResident(resident: Resident): void;
  visitCompany(company: Company): void;
}

abstract class Client {
  constructor(
    protected name: string,
    protected address: string,
    protected addressNumber: number,
  ) {}

  abstract accept(visitor: Visitor): void;
}

class Bank extends Client {
  constructor(
    name: string,
    address: string,
    addressNumber: number,
    readonly branchesInsured: number,
  ) {
    super(name, address, addressNumber);
    this.branchesInsured = branchesInsured;
  }

  accept(visitor: Visitor): void {
    visitor.visitBank(this);
  }
}

class Restaurant extends Client {
  constructor(
    name: string,
    address: string,
    addressNumber: number,
    readonly availableAbroad: boolean,
  ) {
    super(name, address, addressNumber);
    this.availableAbroad = availableAbroad;
  }

  accept(visitor: Visitor): void {
    visitor.visitRestaurant(this);
  }
}

class Resident extends Client {
  constructor(
    name: string,
    address: string,
    addressNumber: number,
    readonly insuranceClass: string,
  ) {
    super(name, address, addressNumber);
    this.insuranceClass = insuranceClass;
  }

  accept(visitor: Visitor): void {
    visitor.visitResident(this);
  }
}

class Company extends Client {
  constructor(
    name: string,
    address: string,
    addressNumber: number,
    readonly nbrOfEmployees: number,
  ) {
    super(name, address, addressNumber);
    this.nbrOfEmployees = nbrOfEmployees;
  }

  accept(visitor: Visitor): void {
    visitor.visitCompany(this);
  }
}

class InsuranceMessagingVisitor implements Visitor {
  sendInsuranceMails(clients: Client[]): void {
    for (const client of clients) {
      client.accept(this);
    }
  }

  visitBank(_bank: Bank): void {
    console.log("Sending mail about theft insurance...");
  }

  visitResident(_resident: Resident): void {
    console.log("Sending mail about medical insurance...");
  }

  visitRestaurant(_restaurant: Restaurant): void {
    console.log("Sending mail about fire and food insurance...");
  }

  visitCompany(_company: Company): void {
    console.log("Sending employees and equipment insurance mail...");
  }
}

const visitor = new InsuranceMessagingVisitor();
const bank = new Bank("Banco", "Rua Avenida", 123, 23);
bank.accept(visitor);
